idaes/apps/grid_integration/multiperiod/price_taker_model.py

```diff
@@ -0,0 +1,429 @@
+# National Technology & Engineering Solutions of Sandia, LLC, Carnegie Mellon
+# University, West Virginia University Research Corporation, et al.
+# All rights reserved.  Please see the files COPYRIGHT.md and LICENSE.md
+# for full copyright and license information.
+#################################################################################
+
+import pandas as pd
+import numpy as np
+
+from pyomo.environ import (
+    ConcreteModel,
+    Block,
+    Var,
+    RangeSet,
+    Objective,
+    Constraint,
+    NonNegativeReals,
+    Expression,
+    maximize,
+)
+
+from idaes.apps.grid_integration import MultiPeriodModel
+from idaes.apps.grid_integration.multiperiod.design_and_operation_models import (
+    DesignModelData,
+    OperationModelData,
+)
+
+from sklearn.cluster import KMeans
+from kneed import KneeLocator
+
+import matplotlib.pyplot as plt
+
+import logging
+
+_logger = logging.getLogger(__name__)
+
+
+class PriceTakerModel(ConcreteModel):
+    def __init__(self, seed=20, horizon_length=24):
+        super().__init__()
+        self._seed = seed
+        self._horizon_length = horizon_length
+
+    @property
+    def seed(self):
+        return self._seed
+
+    @seed.setter
+    def seed(self, value):
+        self._seed = value
+
+    @property
+    def horizon_length(self):
+        return self._horizon_length
+
+    @horizon_length.setter
+    def horizon_length(self, value):
+        if value <= 0:
+            raise ValueError(f"horizon_length must be > 0, but {value} is provided.")
+        self._horizon_length = value
+
+    def generate_daily_data(self, raw_data, day_list):
+
+        daily_data = pd.DataFrame(columns=day_list)
+
+        # Extracting data to populate empty dataframe
+        i = 0
+        j = self._horizon_length
+        day = 1
+        while j <= len(raw_data):
+            daily_data[day] = raw_data[i:j].reset_index(drop=True)
+            i = j
+            j = j + self._horizon_length
+            day = day + 1
+
+        return daily_data
+
+    def reconfigure_raw_data(self, raw_data):
+        """
+        Reconfigures the raw price series data into a usable form
+
+        Args:
+            raw_data: imported price series data
+
+        Returns:
+            daily_data: reconfigured price series data
+        """
+        # Get column headings
+        column_head = raw_data.columns.tolist()
+        # Remove the date/time column
+        scenarios = column_head[2:]
+
+        # Creating an empty dataframe to store daily data for clustering
+        day_list = list(range(1, (len(raw_data) // self._horizon_length) + 1))
+
+        # Generate daily data
+        for i in scenarios[0:1]:
+            daily_data = self.generate_daily_data(
+                raw_data=raw_data[i], day_list=day_list
+            )
+
+        return daily_data, scenarios
+
+    def get_optimal_n_clusters(self, daily_data, kmin=None, kmax=None, plot=False):
```

Comment:
Adam mentioned that we may want to reconsider having a separate function for plotting the elbow plot rather than just an argument in  `get_optimal_n_clusters`. I don't have a strong opinion either way.